distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/DetectEsInstallationTask.java

```diff
@@ -0,0 +1,202 @@
+                    taskInput.setPlugins(detectPluginsFromEsHome(taskInput.getEsHome()));
+                }
+            } else {
+                terminal.println("Did not find a running instance of elasticsearch at " + url);
+                updateTaskInput(taskInput, fetchInfoFromEsSettings(esSettings));
+                taskInput.setPlugins(detectPluginsFromEsHome(taskInput.getEsHome()));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Error detecting existing elasticsearch installation. " + e);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private void updateTaskInput(TaskInput taskInput, Map<?, ?> response) {
+        final Map<String, String> versionMap = (Map<String, String>) response.get("version");
+        if (versionMap != null) {
+            final String vStr = versionMap.get("number");
+            if (vStr != null) {
+                taskInput.setVersion(Version.fromString(vStr));
+            }
+        }
+        taskInput.setNode((String) response.get("name"));
+        taskInput.setCluster((String) response.get("cluster_name"));
+    }
+
+    // package private for unit testing
+    String retrieveUrl(final Settings esSettings) {
+        final int port = Optional.ofNullable(esSettings.get("http.port")).map(this::extractPort).orElse(ES_DEFAULT_PORT);
+        return "http://localhost:" + port;
+    }
+
+    private Integer extractPort(final String port) {
+        try {
+            return Integer.parseInt(port.trim());
+        } catch (Exception ex) {
+            return ES_DEFAULT_PORT;
+        }
+    }
+
+    @SuppressForbidden(reason = "Need to connect to http endpoint for elasticsearch.")
+    private boolean isRunning(final String url) {
+        try {
+            final URL esUrl = new URL(url);
+            final HttpURLConnection conn = (HttpURLConnection) esUrl.openConnection();
+            conn.setRequestMethod("GET");
+            conn.setConnectTimeout(1000);
+            conn.connect();
+            return conn.getResponseCode() == 200;
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    @SuppressForbidden(reason = "Retrieve information on the installation.")
+    private Map<?, ?> fetchInfoFromUrl(final String url) {
+        try {
+            final URL esUrl = new URL(url);
+            final HttpURLConnection conn = (HttpURLConnection) esUrl.openConnection();
+            conn.setRequestMethod("GET");
+            conn.setConnectTimeout(1000);
+            conn.connect();
+
+            final StringBuilder json = new StringBuilder();
+            final Scanner scanner = new Scanner(esUrl.openStream());
+            while (scanner.hasNext()) {
+                json.append(scanner.nextLine());
+            }
+            scanner.close();
+            final ObjectMapper mapper = new ObjectMapper();
+            return mapper.readValue(json.toString(), Map.class);
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)